File: src/audioWfc/constraints/MelodyShapeHardConstraint.java
package audioWfc.constraints;

import audioWfc.wfc.HigherValues;
import audioWfc.musicTheory.OctavedNote;
import audioWfc.wfc.Tile;
import audioWfc.constraints.concepts.HardConstraint;
import audioWfc.constraints.concepts.MelodyShape;
import audioWfc.constraints.concepts.MelodyStep;
import audioWfc.wfc.grabbers.Grabber;

public class MelodyShapeHardConstraint implements HardConstraint<OctavedNote> {
    private Grabber<MelodyShape> grabber;

    public MelodyShapeHardConstraint(Grabber<MelodyShape> grabber){
        this.grabber = grabber;
    }

    @Override
    public boolean check(Tile<OctavedNote> tile, HigherValues higherValues) {
        return checkWithPrev(tile, higherValues) && checkWithNext(tile, higherValues);
    }

    private boolean checkWithPrev(Tile<OctavedNote> tile, HigherValues higherValues) {
        Tile<OctavedNote> prev = tile.getPrev();
        if(!prev.isCollapsed()) return true;

        OctavedNote note1 = tile.getPrev().getValue();
        OctavedNote note2 = tile.getValue();

        return checkConcrete(prev.getPosition(), note1, note2, higherValues);
    }

    private boolean checkWithNext(Tile<OctavedNote> tile, HigherValues higherValues) {
        if(!tile.getNext().isCollapsed()) return true;

        OctavedNote note1 = tile.getValue();
        OctavedNote note2 = tile.getNext().getValue();

        return checkConcrete(tile.getPosition(), note1, note2, higherValues);
    }

    private boolean checkConcrete(int pos, OctavedNote note1, OctavedNote note2, HigherValues higherValues) {
        if(pos > grabber.grab(higherValues).size()) throw new RuntimeException("Melody shape not long enough");

        MelodyStep step = grabber.grab(higherValues).get(pos);

        int distance = note2.MIDIValue() - note1.MIDIValue();

        switch (step){
            case ASCEND -> {return distance > 0;}
            case DESCEND -> {return distance < 0;}
            case STAGNATE -> {return distance == 0;}
            case WILDCARD -> {return true;}
            default -> throw new RuntimeException("Unknown melody step type");
        }
    }

}

File: src/wfc/constraints/MelodyShapeHardConstraint.ts
import { OctavedNote } from "../../music_theory/Note"
import { Grabber } from "../Grabber"
import { HigherValues } from "../HigherValues"
import { Tile } from "../Tile"
import { HardConstraint } from "./concepts/Constraint"
import { MelodyShape } from "./concepts/MelodyShape"
import { noteConstraintTypeToName } from "./constraintUtils"

export const MelodyShapeHardConstraintInit = {
	type: "MelodyShapeHardConstraint" as const,
	shape: [] as MelodyShape,
	validByDefault: true as const,
	reachOverPrev: true,
	reachOverNext: true
}

export type MelodyShapeHardConstraintIR = typeof MelodyShapeHardConstraintInit

export class MelodyShapeHardConstraint implements HardConstraint<OctavedNote> {
	name = noteConstraintTypeToName.get(
		MelodyShapeHardConstraintInit.type,
	)!.name as string
	constructor(private grabber: Grabber<MelodyShape>,
		private reachOverPrev: boolean, private reachOverNext: boolean) {
	}

	private checkPair(
		pos: number,
		first: OctavedNote,
		second: OctavedNote,
		shape: MelodyShape,
	): boolean {
		if (pos >= shape.length) throw new Error("Melody shape not long enough")
		const step = shape[pos]
		const stepSize = second.toMIDIValue() - first.toMIDIValue()

		switch (step) {
			case "ascend":
				return stepSize > 0
			case "descend":
				return stepSize < 0
			case "stagnate":
				return stepSize == 0
			case "wildcard":
				return true
			default:
				throw new Error("Invalid melody shape step")
		}
	}

	check(tile: Tile<OctavedNote>, higherValues: HigherValues): boolean {
		const note = tile.getValue()
		const pos = tile.getPosition()
		const prev = tile.getPrev(this.reachOverPrev)
		const next = tile.getNext(this.reachOverNext)

		const shape = this.grabber(higherValues)

		let out = true
		if (prev.isCollapsed())
			out = out && this.checkPair(pos - 1, prev.getValue(), note, shape)
		if (next.isCollapsed())
			out = out && this.checkPair(pos, note, next.getValue(), shape)
		return out
	}
}
